server/command_list.go

```diff
@@ -28,23 +27,47 @@ func (p *Plugin) runListCommand(args []string, extra *model.CommandArgs) (*model
 }
 
 func (p *Plugin) getUpdatedInstallsForUser(userID string) ([]*Installation, error) {
-	installsForUser, err := p.getInstallationsForUser(userID)
+	pluginInstalls, err := p.getInstallationsForUser(userID)
 	if err != nil {
 		return nil, err
 	}
 
-	var updatedInstall *cloud.Installation
-	for _, install := range installsForUser {
-		updatedInstall, err = p.cloudClient.GetInstallation(install.ID)
-		if err != nil {
-			return nil, errors.Wrapf(err, "could not get updated installation %s", install.ID)
-		}
-		if updatedInstall == nil {
-			return nil, fmt.Errorf("could not find installation %s", install.ID)
+	cloudInstalls, err := p.cloudClient.GetInstallations(&cloud.GetInstallationsRequest{
+		OwnerID:        userID,
+		IncludeDeleted: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: This may become a performance issue as deleted installations in the database start
+	// to grow. We could make it better by assuming that installation is not deleted. If an error
+	// is returned when asked for it, the plugin makes a follow up call to confirm that it was
+	// deleted. If deleted, the installation should be also removed from the plugin storage.
+	for _, cloudInstall := range cloudInstalls {
+		for j, pluginInstall := range pluginInstalls {
+			if cloudInstall.ID == pluginInstall.ID {
+				if cloudInstall.DeleteAt > 0 || cloudInstall.State == cloud.ClusterInstallationStateCreationFailed {
```

Comment:
Lines 49-68 in this PR have a few problems. The first is that `if` statements are nested unnecessarily, when instead `continue` can be used strategically to flatten the routine and make it easier to reason about. Having the "happy path" on the left, achieved by checking error states first and then returning (or in this case `continue`ing) is also [more idiomatic Go](https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88) so I hope that's a style choice we can adhere to. I personally find code where the happy path is down the left easier to reason about.

Lines 58 through 61 also just don't work, see [this Playground link](https://play.golang.org/p/XGTzLuYXIUO). The logical error here is that the last element is always lost.

I've gone ahead and fixed these issues for you, because I was uncertain about my own concerns and wanted to verify before leaving this review. If we don't have a unit test for this method, we should probably write one before merging this to catch issues like the one on lines 58-61. 
```go
	for _, cloudInstall := range cloudInstalls {
		for j, pluginInstall := range pluginInstalls {
			if cloudInstall.ID != pluginInstall.ID {
				continue
			}

			if cloudInstall.DeleteAt > 0 || cloudInstall.State == cloud.ClusterInstallationStateCreationFailed {
				err = p.deleteInstallation(pluginInstall.ID)
				if err != nil {
					p.API.LogError(err.Error(), pluginInstall.ID)
					continue
				}
				// Notify the user that installation was removed.
				p.PostBotDM(userID, fmt.Sprintf("Cloud installation ID %s has been removed from your Mattermost app.", pluginInstall.ID))
				// Remove key from the plugin installations.
				pluginInstalls = append(pluginInstalls[:j], pluginInstalls[j+1:]...)
			} else {
				pluginInstall.Installation = *cloudInstall
				pluginInstall.License = "hidden"
			}

			break
		}
	}
```
I have only tested the line with append(), which you can see in the Playground link. The rest of the `if` statement flattening that I've done here should be tested.

Reply:
Good catch, thanks @gigawhitlocks! I think I know what happened, but it definitely needs more testing. I have to go back to some old changes but I'm pretty sure that I commit some old code. I think that L59 shouldn't be there. If removed, the last element takes the place of the one being deleted. The whole idea is to avoid reallocations by using append function since I didn't know the since of these arrays. I will look at the rest of your comments and add more tests later, but it should work if L59 (which shouldn't be there at first place) is deleted.

Reply:
Ah, I thought using `append` with a slice was smarter than it is, but it looks like you are right and `append` does copy the elements around. Your optimization, once the logic is corrected, will be more performant for sure, and I'm more than happy to see that checked in. I'd suggest creating a helper function for those 4 lines though, because then it should be trivial and worthwhile to write a unit test for the "remove element" function, since it is a little tricky.

Reply:
Could you address some of the style issues in this block that I brought up, besides the bug fixed in commit 02c0087? This block of code is difficult to understand and I provided some refactored code to make it simpler. 

Regarding the bug fixed in 02c0087, I am concerned that changing the length of a list that Go is currently iterating over with `for ... range` will cause an index out of bounds exception. Is there a test for this code block? Considering its complexity and that we have found a bug in it already, I think this block of code needs to be refactored and the surrounding function should be exercised and protected with a test before it is merged.

See also this Playground link to see the behavior of removing an element from a list while iterating over it with `range`:
https://play.golang.org/p/wJrY8BLbtrl

Reply:
Sorry, that's a problem caused by some immutability nature of `range`. Here is the fixed version:
https://play.golang.org/p/NxRHcVT3P8N 

I'm looking at the other parts of the comment.